client/keys/add.go

```diff
@@ -111,17 +139,99 @@ func runAddCmd(cmd *cobra.Command, args []string) error {
 		// print out results without the seed phrase
 		viper.Set(flagNoBackup, true)
 		printCreate(info, "")
-	} else {
-		algo := keys.SigningAlgo(viper.GetString(flagType))
-		info, seed, err := kb.CreateMnemonic(name, keys.English, pass, algo)
+		return nil
+	}
+
+	if !useDefaults {
+		mnemonic, err = client.GetString("Enter your bip39 mnemonic, or hit enter to generate one.", buf)
+		if err != nil {
+			return err
+		}
+	}
+
+	if len(mnemonic) == 0 {
+		// read entropy seed straight from crypto.Rand and convert to mnemonic
+		entropySeed, err := bip39.NewEntropy(mnemonicEntropySize)
 		if err != nil {
 			return err
 		}
-		printCreate(info, seed)
+
+		mnemonic, err = bip39.NewMnemonic(entropySeed[:])
+		if err != nil {
+			return err
+		}
+	}
+
+	// get bip39 passphrase
+	var bip39Passphrase string
```

Comment:
Whats the difference between the `bip39Passphrase` and the `encryptPassword`?

Reply:
From ` cosmos-sdk/crypto/keys/types.go`:

```go
	// Compute a BIP39 seed from th mnemonic and bip39Passwd.
	// Derive private key from the seed using the BIP44 params.
	// Encrypt the key to disk using encryptPasswd.
	// See https://github.com/cosmos/cosmos-sdk/issues/2095
	Derive(name, mnemonic, bip39Passwd,
		encryptPasswd string, params hd.BIP44Params) (Info, error)
```

Reply:
bip39passphrase is part of the derivation of the root key from the mnemonic. encryptPassword is for encrypting the derived private key (for the given bip44 path) on the disk

Reply:
Great explanation 👌